wgpu-rs/src/lib.rs

```diff
@@ -465,53 +465,65 @@ pub enum BufferMapAsyncResult<T> {
     Error,
 }
 
-struct BufferMapReadAsyncUserData<F: FnOnce(BufferMapAsyncResult<&[u8]>)> {
+struct BufferMapReadAsyncUserData<T,F> 
+    where F: FnOnce(BufferMapAsyncResult<&[T]>) {
     size: u32,
     callback: F,
+    phantom: std::marker::PhantomData<T>,
 }
 
-struct BufferMapWriteAsyncUserData<F: FnOnce(BufferMapAsyncResult<&mut [u8]>)> {
+struct BufferMapWriteAsyncUserData<T, F>
+    where F: FnOnce(BufferMapAsyncResult<&mut [T]>) {
     size: u32,
     callback: F,
+    phantom: std::marker::PhantomData<T>,
 }
 
 impl Buffer {
     pub fn set_sub_data(&self, offset: u32, data: &[u8]) {
         wgn::wgpu_buffer_set_sub_data(self.id, offset, data.len() as u32, data.as_ptr());
     }
 
-    pub fn map_read_async<F>(&self, start: u32, size: u32, callback: F)
-            where F: FnOnce(BufferMapAsyncResult<&[u8]>) {
-        extern "C" fn buffer_map_read_callback_wrapper<F>(status: wgn::BufferMapAsyncStatus, data: *const u8, userdata: *mut u8)
-                where F: FnOnce(BufferMapAsyncResult<&[u8]>) {
-            let userdata = unsafe { Box::from_raw(userdata as *mut BufferMapReadAsyncUserData<F>) };
-            let data = unsafe { slice::from_raw_parts(data, userdata.size as usize) };
+    pub fn map_read_async<T, F>(&self, start: u32, size: u32, callback: F)
```

Comment:
The trickiest thing here is to establish a safe contract for `T`. Remember that wgpu-rs is a safe API!
For example, what happens if the user puts `String` there? What if `T = &U` ? and so forth.
In gfx-rs for this very reason we have [trait Pod](https://github.com/gfx-rs/gfx/blob/60fb10082bf8a6b7b2b1de06b0980bf83794368a/src/hal/src/memory.rs#L14), but it wasn't ideal either, see https://github.com/gfx-rs/gfx/issues/2573

Reply:
Just in case, going to call in @Gankro here, in case they faced a similar situation with the standard library somehow. TL;DR: we are trying to have a safe API that gets user structures and writes them into byte buffers for GPU.

Reply:
Is`Copy` a sufficiently restrictive trait bound? I can't come up with a simple example where I'd need to transfer a non-copyable type in a buffer, and it'd rule out obviously problematic cases like `T = String` and `T = &mut U`.